### PR TITLE
Add some local caching for the vocabs as well as multiple attempts to get them

### PR DESCRIPTION
**Pull request by somexpert**

### CODE REVIEW

**Review by ldko — COMMENTED** [review]
Looks good to me. Just a couple clean up suggestions.

File: pyuntl/untl_structure.py [hunk 1]
```diff
@@ -320,19 +324,30 @@ def create_form_groupings(self,
         element_list.sort(key=lambda obj: sort_order.index(obj.group_name))
         return element_list
```
[the remark this answers is not in the document]
Comment (somexpert) [reply pyuntl/untl_structure.py:326]: will do!

File: pyuntl/untl_structure.py [hunk 2]
```diff
@@ -320,19 +324,30 @@ def create_form_groupings(self,
...
-        try:
-            vocab_data = json.loads(urllib.request.urlopen(vocab_url).read())
-        except:
-            raise UNTLStructureException('Could not retrieve the vocabularies')
-        return vocab_data
+def get_vocabularies():
+    """Get the vocabularies to pull the qualifiers from."""
+    # Timeout in seconds.
+    timeout = 5
+    socket.setdefaulttimeout(timeout)
```
[the remark this answers is not in the document]
Comment (somexpert) [reply pyuntl/untl_structure.py:331]: oh, definitely like that, I'll make the change

[the remark this answers is not in the document]
Comment (ldko) [reply pyuntl/untl_structure.py]: and no more need to `import socket?

[the remark this answers is not in the document]
Comment (somexpert) [reply pyuntl/untl_structure.py]: ah, of course I forgot that. I'll fix it!

**Review by ldko — APPROVED** [review]
Looks good to me! :+1:

**Comment by gracieflores** [comment]
Looks good!

**Comment by gracieflores** [comment]
Can one of you help me understand what the best solution is for the rdflib package issue. Do we want to update the version of it in the setup file and fix the one test that was affected by the new version or do we want to keep the test as is and make sure that setup doesn't allow for the newer version of rdflib?

Here is the line of code in the test that fails when using rdflib 6.1.1
https://github.com/unt-libraries/pyuntl/blob/resilient-vocabs/tests/untldoc_test.py#L554

The test passes if we change that line to:
rdf_lines = sorted(rdf_xml.split())

From reading the docs we would only need to stick to the older versions if we needed Python 2 support.

File: pyuntl/untldoc.py [hunk 3]
```diff
@@ -578,19 +579,17 @@ def generate_rdf_xml(dc_dict):
...
 
 
 def retrieve_vocab():
     """Get the vocabulary data for the field.
 
     Retrieves and returns the vocabulary. Upon failure, returns None.
     """
-    url = VOCABULARIES_URL.replace('all', 'all-verbose')
     try:
-        vocab_data = urllib.request.urlopen(url).read()
```
Comment (ldko) [inline pyuntl/untldoc.py:591]: Didn't catch this before, but running flake8 now, I think with this change we no longer need to import `urllib.request` in this file.

File: pyuntl/untl_structure.py [hunk 4]
```diff
@@ -321,18 +324,32 @@ def create_form_groupings(self,
...
+    """Get the vocabularies to pull the qualifiers from."""
+    # Create the ordered vocabulary URL.
+    vocab_url = VOCABULARIES_URL.replace('all', 'all-verbose')
+    # Try to get the cached vocabs, only hitting the live vocabs when needed
+    if vocab_url not in VOCAB_CACHE:
+        # Try to retrieve the fresh vocabs up to 3 times in case there are availability issues
+        attempt = 0
+        while True:
+            try:
+                VOCAB_CACHE[vocab_url] =  json.loads(
```
Comment (ldko) [inline pyuntl/untl_structure.py:341]: Via flake8 check, there is an extra space after the `=` here.

**Comment by gracieflores** [comment]
@ldko @somexpert I pushed a few changes and it is ready for review.

File: tests/untl_structure_test.py [hunk 5]
```diff
@@ -372,26 +373,40 @@ def test_FormGenerator_adjustable_items(_):
...
+    assert mock_urlopen.call_count == 0
+    assert vocabularies == 'name'
+
+
+@patch('urllib.request.urlopen', side_effect=[Exception, io.StringIO(json.dumps(VOCAB))])
+def test_get_vocabularies_multiple_attempts_succeeds(mock_urlopen, capsys):
+    """Test vocabularies are fetched when not in cache."""
+    output = 'Exception caught while trying to retrieve vocabs: '
+    us.VOCAB_CACHE = {}
+    vocabulaires = us.get_vocabularies()
```
Comment (ldko) [inline tests/untl_structure_test.py:406]: Doesn't affect outcomes, but we could fix the typo on the variable name "vocabulaires".

**Review by ldko — APPROVED** [review]
I just have one minor comment about a typo, otherwise this looks good to me!

**Review by somexpert — COMMENTED** [review]
This looks good to me! :thumbsup: @gracieflores I like how you did the test for using the vocab cache, in my mind I was going a different way with it, but yours is easy to read and gets to the point better than what I had in my mind :smile:

**Comment by somexpert** [comment]
Oh! _One_ small thing: can we get just _one_ more test to show that we successfully put vocabs _into_ the cache when it's empty and we retrieve something?

**Comment by somexpert** [comment]
Thanks for adding that check! I'm good with this getting merged now :thumbsup:

**Comment by gracieflores** [comment]
Should I have used `vocab_url = VOCABULARIES_URL.replace('all', 'all-verbose')` instead of hard coding in the url in case that ever changes?

**Comment by somexpert** [comment]
Oh, actually, yeah, that would be good :smile:

**Comment by ldko** [comment]
I think that is a good idea to do...not hard code the URL.

**Comment by ldko** [comment]
Ok, I think it can be merged. :partying_face:
